In [ ]:
import pandas as pd
from datasets import load_dataset

ds_space_weather_events = load_dataset("juliensimon/donki-space-weather-events", split="all")
ds_omni_solar_wind_parameters = load_dataset("juliensimon/omni-solar-wind-parameters", split="all")
ds_solar_wind = load_dataset("juliensimon/solar-wind", split="all")
ds_space_weather_indices = load_dataset("juliensimon/space-weather-indices", split="all")

df_space_weather_events = ds_space_weather_events.to_pandas()
df_omni_solar_wind_parameters = ds_omni_solar_wind_parameters.to_pandas()
df_solar_wind = ds_solar_wind.to_pandas()
df_space_weather_indices = ds_space_weather_indices.to_pandas()

# Flares come from the team-cleaned parquet (see exploratory-data-analysis/solar-flare-events/eda.ipynb):
# garbage rows dropped, duration_minutes + log_peak_flux + end_time_imputed added.
# Raw OMNI is kept (not the cleaned parquet) because we need ap_index_nt for the label
# and NaN-tolerant window aggregation instead of complete-case row dropping.
df_solar_flare_events = pd.read_parquet('data/clean-data/solar-flare-events.parquet')

/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


In [2]:
df_space_weather_events.head()

,event_type,activity_id,start_time,source_location,active_region,note,link,cme_speed_kms,cme_half_angle_deg,cme_latitude,cme_longitude,cme_type,cme_time_21_5,cme_measurement,linked_events,gst_max_kp,gst_kp_count
0,IPS,2010-01-20T20:20:00-IPS-001,2010-01-20 20:20:00+00:00,NaN,NaN,STEREO B,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/IP...,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
1,IPS,2010-02-05T03:33:00-IPS-001,2010-02-05 03:33:00+00:00,NaN,NaN,STEREO A,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/IP...,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
2,IPS,2010-02-14T07:50:00-IPS-001,2010-02-14 07:50:00+00:00,NaN,NaN,STEREO B,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/IP...,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
3,IPS,2010-03-05T18:25:00-IPS-001,2010-03-05 18:25:00+00:00,NaN,NaN,STEREO A,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/IP...,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
4,CME,2010-04-03T09:54:00-CME-001,2010-04-03 09:54:00+00:00,S20E05,NaN,SDO images were unavailable. CME source locati...,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/CM...,620.0,26.0,7.0,8.0,C,2010-04-03 17:16:00+00:00,null,"2010-04-03T09:04:00-FLR-001, 2010-04-05T07:54:...",NaN,NaN


In [3]:
df_space_weather_events.head()

,event_type,activity_id,start_time,source_location,active_region,note,link,cme_speed_kms,cme_half_angle_deg,cme_latitude,cme_longitude,cme_type,cme_time_21_5,cme_measurement,linked_events,gst_max_kp,gst_kp_count
0,IPS,2010-01-20T20:20:00-IPS-001,2010-01-20 20:20:00+00:00,NaN,NaN,STEREO B,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/IP...,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
1,IPS,2010-02-05T03:33:00-IPS-001,2010-02-05 03:33:00+00:00,NaN,NaN,STEREO A,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/IP...,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
2,IPS,2010-02-14T07:50:00-IPS-001,2010-02-14 07:50:00+00:00,NaN,NaN,STEREO B,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/IP...,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
3,IPS,2010-03-05T18:25:00-IPS-001,2010-03-05 18:25:00+00:00,NaN,NaN,STEREO A,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/IP...,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
4,CME,2010-04-03T09:54:00-CME-001,2010-04-03 09:54:00+00:00,S20E05,NaN,SDO images were unavailable. CME source locati...,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/CM...,620.0,26.0,7.0,8.0,C,2010-04-03 17:16:00+00:00,null,"2010-04-03T09:04:00-FLR-001, 2010-04-05T07:54:...",NaN,NaN


In [4]:
df_space_weather_events[df_space_weather_events['start_time'] > '2017-02-19'].head()

,event_type,activity_id,start_time,source_location,active_region,note,link,cme_speed_kms,cme_half_angle_deg,cme_latitude,cme_longitude,cme_type,cme_time_21_5,cme_measurement,linked_events,gst_max_kp,gst_kp_count
3561,HSS,2017-02-19T10:00:00-HSS-001,2017-02-19 10:00:00+00:00,NaN,NaN,NaN,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/HS...,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
3562,CME,2017-02-19T12:30:00-CME-001,2017-02-19 12:30:00+00:00,N10E01,NaN,Source is an eruption visible around 05:31 in ...,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/CM...,450.0,42.0,11.0,12.0,S,2017-02-19 19:02:00+00:00,null,NaN,NaN,NaN
3563,CME,2017-02-20T19:12:00-CME-001,2017-02-20 19:12:00+00:00,NaN,NaN,Source is large eruption visible off the NE li...,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/CM...,155.0,11.0,29.0,-90.0,S,2017-02-21 15:21:00+00:00,null,NaN,NaN,NaN
3564,HSS,2017-02-23T05:00:00-HSS-001,2017-02-23 05:00:00+00:00,NaN,NaN,NaN,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/HS...,NaN,NaN,NaN,NaN,NaN,NaT,NaN,NaN,NaN,NaN
3565,CME,2017-02-23T11:09:00-CME-001,2017-02-23 11:09:00+00:00,NaN,NaN,Source is opening field lines seen off the lim...,https://kauai.ccmc.gsfc.nasa.gov/DONKI/view/CM...,675.0,20.0,29.0,-44.0,C,2017-02-23 16:41:00+00:00,null,NaN,NaN,NaN


In [5]:
df_omni_solar_wind_parameters[df_omni_solar_wind_parameters['datetime'] > '2017-02-19']["b_magnitude_avg_nt"].isnull().sum() / len(df_omni_solar_wind_parameters[df_omni_solar_wind_parameters['datetime'] > '2017-02-19'])

np.float64(0.06076025488314001)

In [6]:
df_solar_flare_events.head()

,satellite,start_time,peak_time,peak_flux_wm2,goes_class,end_time,goes_class_letter,duration_minutes,end_time_imputed,log_peak_flux,duration_outlier,flux_outlier
0,GOES-16,2017-02-09 00:41:00,2017-02-09 00:50:00,3.711663e-07,B3.7,2017-02-09 00:56:00,B,15.0,0,-6.430431,False,False
1,GOES-16,2017-02-09 01:30:00,2017-02-09 01:40:00,4.310535e-07,B4.3,2017-02-09 01:42:00,B,12.0,0,-6.365469,False,False
2,GOES-16,2017-02-09 01:45:00,2017-02-09 01:51:00,1.694939e-06,C1.6,2017-02-09 01:55:00,C,10.0,0,-5.770846,False,False
3,GOES-16,2017-02-09 02:31:00,2017-02-09 02:40:00,3.234027e-07,B3.2,2017-02-09 02:48:00,B,17.0,0,-6.490256,False,False
4,GOES-16,2017-02-09 02:55:00,2017-02-09 02:59:00,3.839661e-07,B3.8,2017-02-09 03:08:00,B,13.0,1,-6.415707,False,False


In [7]:
df_solar_wind.head()

,time_tag,density,speed,temperature,bt,bx_gsm,by_gsm,bz_gsm
0,2026-03-17 08:31:00,0.38,467.8,57895.0,4.49,-3.47,2.74,-0.79
1,2026-03-17 08:32:00,0.41,471.6,71499.0,4.58,-3.38,3.08,-0.21
2,2026-03-17 08:33:00,0.42,472.1,76974.0,4.39,-3.21,2.99,-0.10
3,2026-03-17 08:34:00,0.45,465.1,51641.0,4.27,-3.45,2.47,-0.47
4,2026-03-17 08:35:00,0.95,489.6,103878.0,4.39,-3.49,2.63,-0.33


In [8]:
df_space_weather_indices.head()

,date,bartels_rotation,bartels_day,kp_0000,kp_0300,kp_0600,kp_0900,kp_1200,kp_1500,kp_1800,...,f107_obs,f107_adj,f107_data_type,f107_obs_center81,f107_obs_last81,f107_adj_center81,f107_adj_last81,is_storm,storm_level,data_type
0,1957-10-01,1700,19,43.0,40.0,30.0,20.0,37.0,23.0,43.0,...,269.3,269.8,OBS,266.6,230.9,266.8,235.5,False,G5,observed
1,1957-10-02,1700,20,37.0,37.0,17.0,17.0,27.0,23.0,17.0,...,253.3,253.6,OBS,267.4,231.7,267.5,236.2,False,G5,observed
2,1957-10-03,1700,21,27.0,20.0,13.0,33.0,37.0,47.0,43.0,...,266.3,266.4,OBS,268.1,232.7,268.1,237.1,False,G5,observed
3,1957-10-04,1700,22,30.0,30.0,23.0,27.0,23.0,27.0,30.0,...,238.2,238.2,OBS,268.8,233.3,268.7,237.7,False,G5,observed
4,1957-10-05,1700,23,30.0,30.0,17.0,23.0,20.0,27.0,27.0,...,246.2,246.0,OBS,269.3,233.9,269.1,238.1,False,G5,observed


In [9]:
kp_cols = [c for c in df_space_weather_indices.columns
           if c.startswith("kp_") and c != "kp_sum"]

df_storms = pd.DataFrame({
    "date": df_space_weather_indices["date"],
    "kp_max": df_space_weather_indices[kp_cols].max(axis=1)
})

def storm_level(kp_max):
    if kp_max >= 90:
        return "G5"
    elif kp_max >= 80:
        return "G4"
    elif kp_max >= 70:
        return "G3"
    elif kp_max >= 60:
        return "G2"
    elif kp_max >= 50:
        return "G1"
    return None

df_storms["storm_level"] = df_storms["kp_max"].apply(storm_level)
df_storms["is_storm"] = df_storms["kp_max"] >= 50

df_storms["storm_level"].value_counts().sum()

np.int64(3769)

## Dataset Join Strategy — CME-Anchored

**One row = one CME event.** Each CME is a discrete "will this hit Earth and cause a storm?" question. **Decision time is launch + 12h** (the end of the wind feature window), giving 0.5–3.5 days of effective warning.

**Datasets used:**
- `space_weather_events` (CME rows) — anchor: one row per CME, with speed, half-angle, latitude, longitude, type
- `data/clean-data/solar-flare-events.parquet` — team-cleaned GOES flares, matched by time proximity (nearest flare ≤ 6h before CME)
- `omni_solar_wind_parameters` (raw) — solar wind around launch (−24h → +12h) via independent per-CME searchsorted windows; also provides Kp for the label
- `space_weather_indices` — not used directly (Kp comes from omni)

**Dropped:**
- `solar_wind` — 2026 real-time data only
- Day-based aggregation — misaligns causes (flare/CME at T) with effects (storm at T+1 to T+4)

**Time window:** 2017-02-09 – 2024-12-31 (flare catalog starts 2017-02-09; earlier CMEs would have structurally missing flare features)

**Feature groups per CME row:**
1. CME intrinsic properties: speed, half-angle, latitude, |longitude|, type
2. Associated flare (if any ≤ 6h prior): GOES class, log peak flux, duration, has_flare flag
3. Solar wind, launch − 24h → launch + 12h: mean/min Bz, southward-Bz fractions, mean |B|, mean/max speed, density, pressure, electric field, temperature
4. Solar-cycle context in the same window: F10.7, sunspot number

**Targets:**
- `kp_max_post` — max Kp in the 1–4 days after CME start (the typical CME travel window)
- `storm` = 1 if `kp_max_post` ≥ 5 (NOAA G1 threshold)
- `storm_attributed` — same, but each storm episode is credited only to the fastest/widest candidate CME (reduces multi-CME label noise)

In [10]:
import pandas as pd

print("=== omni columns ===")
print(df_omni_solar_wind_parameters.columns.tolist())
print("\n=== space_weather_indices columns ===")
print(df_space_weather_indices.columns.tolist())
print("\n=== solar_flare_events date range ===")
print(df_solar_flare_events['start_time'].min(), "→", df_solar_flare_events['start_time'].max())
print("\n=== space_weather_events event_type counts ===")
print(df_space_weather_events['event_type'].value_counts())

=== omni columns ===
['datetime', 'bartels_rotation_number', 'b_magnitude_avg_nt', 'b_magnitude_vector_nt', 'b_lat_angle_gse_deg', 'b_lon_angle_gse_deg', 'bx_gse_nt', 'by_gse_nt', 'bz_gse_nt', 'by_gsm_nt', 'bz_gsm_nt', 'sigma_b_magnitude_nt', 'sigma_b_vector_nt', 'sigma_bx_nt', 'sigma_by_nt', 'sigma_bz_nt', 'proton_temperature_k', 'proton_density_cm3', 'flow_speed_kms', 'flow_lon_angle_deg', 'flow_lat_angle_deg', 'alpha_proton_ratio', 'flow_pressure_npa', 'sigma_t_k', 'sigma_n_cm3', 'sigma_v_kms', 'sigma_phi_v_deg', 'sigma_theta_v_deg', 'sigma_alpha_proton_ratio', 'electric_field_mvpm', 'plasma_beta', 'alfven_mach_number', 'kp_index', 'sunspot_number', 'dst_index_nt', 'ae_index_nt', 'proton_flux_gt1mev', 'proton_flux_gt2mev', 'proton_flux_gt4mev', 'proton_flux_gt10mev', 'proton_flux_gt30mev', 'proton_flux_gt60mev', 'ap_index_nt', 'f107_index_sfu', 'pc_n_index', 'al_index_nt', 'au_index_nt', 'magnetosonic_mach_number']

=== space_weather_indices columns ===
['date', 'bartels_rotation', 

In [11]:
# Start at the GOES flare catalog start so "no flare matched" always means
# "genuinely no flare", never "flare data not available yet". Using 2010 would
# entangle flare-missingness with the solar cycle and the train/test split.
TRAIN_START = '2017-02-09'   # solar_flare_events coverage begins here
TRAIN_END   = '2024-12-31'

### 1. CME anchor table

In [12]:
import pandas as pd

df_cme = df_space_weather_events[df_space_weather_events['event_type'] == 'CME'].copy()

# normalize timestamps to tz-naive UTC
df_cme['cme_time'] = pd.to_datetime(df_cme['start_time'], utc=True).dt.tz_localize(None)
df_cme = df_cme[
    (df_cme['cme_time'] >= TRAIN_START) & (df_cme['cme_time'] <= TRAIN_END)
].reset_index(drop=True)

# encode CME type: S=1 (slow/narrow), C=2 (common), O=3, R=4 (fast/halo)
cme_type_map = {'S': 1, 'C': 2, 'O': 3, 'R': 4}
df_cme['cme_type_num'] = df_cme['cme_type'].map(cme_type_map)

df_cme = df_cme[[
    'cme_time', 'cme_speed_kms', 'cme_half_angle_deg',
    'cme_latitude', 'cme_longitude', 'cme_type_num',
]].copy()

df_cme['cme_id'] = range(len(df_cme))

print(f"CME events 2017–2024: {len(df_cme)}")
df_cme.head()

CME events 2017–2024: 4737


,cme_time,cme_speed_kms,cme_half_angle_deg,cme_latitude,cme_longitude,cme_type_num,cme_id
0,2017-02-10 13:48:00,594.0,10.0,18.0,148.0,2.0,0
1,2017-02-11 06:00:00,456.0,18.0,-1.0,93.0,1.0,1
2,2017-02-15 03:09:00,405.0,12.0,16.0,-173.0,1.0,2
3,2017-02-18 02:30:00,239.0,17.0,26.0,-110.0,1.0,3
4,2017-02-18 07:24:00,492.0,27.0,4.0,88.0,1.0,4


### 2. Match associated flare (time proximity)

DONKI activity IDs in `linked_events` don't map to our GOES-based `solar_flare_events`, so we match by time: find the most energetic flare that peaked within 6 hours before each CME. This captures the eruption that likely triggered it.

In [13]:
class_map = {'A': 0, 'B': 1, 'C': 2, 'M': 3, 'X': 4}

df_flares = df_solar_flare_events.copy()
df_flares['peak_time'] = pd.to_datetime(df_flares['peak_time'])
df_flares = df_flares.dropna(subset=['peak_time'])  # merge_asof requires no nulls on key
df_flares['flare_class_num'] = df_flares['goes_class_letter'].map(class_map)
df_flares = df_flares.sort_values('peak_time').reset_index(drop=True)

df_cme_sorted = df_cme.sort_values('cme_time').reset_index(drop=True)

# For each CME, find the nearest flare peak that occurred ≤ 6h before it.
# merge_asof matches each CME to the most recent flare at or before cme_time.
FLARE_FEATURE_COLS = ['goes_class_letter', 'flare_class_num', 'log_peak_flux', 'duration_minutes', 'end_time_imputed']
flare_matched = pd.merge_asof(
    df_cme_sorted,
    df_flares[['peak_time'] + FLARE_FEATURE_COLS],
    left_on='cme_time',
    right_on='peak_time',
    direction='backward',
    tolerance=pd.Timedelta('6h'),
)

flare_matched = flare_matched.rename(columns={
    'goes_class_letter': 'flare_class',
    'log_peak_flux':     'flare_log_peak_flux',
    'duration_minutes':  'flare_duration_min',
    'end_time_imputed':  'flare_end_time_imputed',
})
flare_matched = flare_matched.drop(columns=['peak_time'])

# Explicit indicator instead of zero-filling flare features: within 2017+,
# NaN flare features genuinely mean "no flare in the 6h window", and the
# indicator lets models separate "no flare" from any imputed feature values.
flare_matched['has_flare'] = flare_matched['flare_class'].notna().astype(int)

matched = flare_matched['has_flare'].sum()
print(f"CMEs with an associated flare (≤6h prior): {matched} / {len(flare_matched)}  ({matched/len(flare_matched):.1%})")
flare_matched.head()

CMEs with an associated flare (≤6h prior): 3773 / 4737  (79.6%)


,cme_time,cme_speed_kms,cme_half_angle_deg,cme_latitude,cme_longitude,cme_type_num,cme_id,flare_class,flare_class_num,flare_log_peak_flux,flare_duration_min,flare_end_time_imputed,has_flare
0,2017-02-10 13:48:00,594.0,10.0,18.0,148.0,2.0,0,<NA>,NaN,NaN,NaN,NaN,0
1,2017-02-11 06:00:00,456.0,18.0,-1.0,93.0,1.0,1,<NA>,NaN,NaN,NaN,NaN,0
2,2017-02-15 03:09:00,405.0,12.0,16.0,-173.0,1.0,2,<NA>,NaN,NaN,NaN,NaN,0
3,2017-02-18 02:30:00,239.0,17.0,26.0,-110.0,1.0,3,B,1.0,-6.885685,31.0,0.0,1
4,2017-02-18 07:24:00,492.0,27.0,4.0,88.0,1.0,4,<NA>,NaN,NaN,NaN,NaN,0


### 3. Solar wind background (launch − 24h → launch + 12h)

These features capture the ambient interplanetary conditions around the CME's launch, plus solar-cycle context (F10.7, sunspots). Background Bz, speed, and density all influence how geoeffective the CME will be when it arrives — and the 12h *after* launch can catch earlier disturbances already in transit toward Earth.

**Why +12h is safe (no leakage):** the CME takes ≥ ~1 day to reach the L1 monitors, so wind readings 12h after launch are still ambient background, not the CME's own signature — and the label window starts at launch + 1 day, comfortably after the feature window ends. The honest cost is that the prediction is now issued 12h after launch, so effective warning is 0.5–3.5 days instead of 1–4. Extending to +12h improved attributed-label classification (XGB PR-AUC 0.106 → 0.156, ROC-AUC 0.835 → 0.860) with no change on the naive label.

Implementation note: 83% of CMEs launch within 24h of the previous one (median gap ≈ 7.6h), so windows overlap heavily. A forward `merge_asof` would assign each OMNI reading to only the *next* CME, silently truncating most windows to the inter-CME gap. Instead, `searchsorted` gives every CME its own full, independent window; the same OMNI hour can contribute to several CMEs.

In [14]:
import numpy as np

df_omni = df_omni_solar_wind_parameters.copy()
df_omni['datetime'] = pd.to_datetime(df_omni['datetime'])
df_omni = df_omni[
    (df_omni['datetime'] >= TRAIN_START) & (df_omni['datetime'] <= TRAIN_END)
].sort_values('datetime').reset_index(drop=True)

# Strong keeps per the OMNI EDA, plus its two "optional" features
# (electric_field_mvpm, proton_temperature_k) and solar-cycle context.
WIND_COLS = [
    'bz_gsm_nt', 'b_magnitude_avg_nt', 'flow_speed_kms',
    'proton_density_cm3', 'flow_pressure_npa',
    'electric_field_mvpm', 'proton_temperature_k',
    'f107_index_sfu', 'sunspot_number',
]

# Window: [launch − 24h, launch + 12h]. The +12h stays well clear of the
# label window (starts at +1 day) and of the CME's own L1 arrival (≥ ~1 day),
# so it adds signal (earlier disturbances in transit) without leakage.
# Decision time is therefore launch + 12h.
WINDOW_PRE_H  = 24
WINDOW_POST_H = 12

omni_times = df_omni['datetime'].values
w_starts = (flare_matched['cme_time'] - pd.Timedelta(hours=WINDOW_PRE_H)).values
w_ends   = (flare_matched['cme_time'] + pd.Timedelta(hours=WINDOW_POST_H)).values
idx_starts = np.searchsorted(omni_times, w_starts)
idx_ends   = np.searchsorted(omni_times, w_ends, side='right')

def window_stats(s, e):
    if e <= s:
        return {}
    w = df_omni.iloc[s:e]
    bz = w['bz_gsm_nt']
    return {
        'wind_bz_mean':        bz.mean(),
        'wind_bz_min':         bz.min(),
        # sustained southward Bz matters more than single readings (solar-wind EDA)
        'wind_bz_south_frac':  (bz < 0).mean() if bz.notna().any() else np.nan,
        'wind_bz_strong_south_frac': (bz < -10).mean() if bz.notna().any() else np.nan,
        'wind_imf_mean':       w['b_magnitude_avg_nt'].mean(),
        'wind_speed_mean':     w['flow_speed_kms'].mean(),
        'wind_speed_max':      w['flow_speed_kms'].max(),
        'wind_density_mean':   w['proton_density_cm3'].mean(),
        'wind_pressure_mean':  w['flow_pressure_npa'].mean(),
        'wind_efield_mean':    w['electric_field_mvpm'].mean(),
        'wind_temp_mean':      w['proton_temperature_k'].mean(),
        'ctx_f107':            w['f107_index_sfu'].mean(),
        'ctx_sunspots':        w['sunspot_number'].mean(),
    }

df_wind_pre = pd.DataFrame(
    [window_stats(s, e) for s, e in zip(idx_starts, idx_ends)],
    index=flare_matched['cme_id'],
).reset_index()

with_wind = df_wind_pre['wind_bz_mean'].notna().sum()
print(f"CMEs with launch-window wind data: {with_wind} / {len(flare_matched)}")
df_wind_pre.head()

CMEs with launch-window wind data: 4737 / 4737


,cme_id,wind_bz_mean,wind_bz_min,wind_bz_south_frac,wind_bz_strong_south_frac,wind_imf_mean,wind_speed_mean,wind_speed_max,wind_density_mean,wind_pressure_mean,wind_efield_mean,wind_temp_mean,ctx_f107,ctx_sunspots
0,0,-0.569444,-3.7,0.611111,0.0,4.672222,421.194444,445.0,5.500000,2.005278,0.243889,70931.944444,72.027778,22.555556
1,1,-0.637838,-3.2,0.621622,0.0,3.610811,411.972973,440.0,4.745946,1.616486,0.265946,56423.216216,72.972973,23.486486
2,2,-0.030556,-3.3,0.388889,0.0,2.861111,307.750000,322.0,12.597222,2.053611,0.010000,18074.472222,72.833333,19.555556
3,3,-1.097222,-6.3,0.611111,0.0,6.061111,495.444444,536.0,6.983333,3.387500,0.509722,149947.583333,73.675000,13.583333
4,4,-0.600000,-6.3,0.472222,0.0,6.077778,500.361111,536.0,6.444444,3.187778,0.266111,137876.500000,73.966667,13.444444


### 4. Storm label — Kp-based, with per-storm attribution

For each CME, take the max Kp in the 1–4 day window after launch (typical Sun→Earth travel time; day 0 is skipped to avoid contamination from launch-time disturbances). `storm = 1` if max Kp ≥ 5 — the NOAA G1 threshold, and the same definition the OMNI EDA used (`kp_index` is stored as Kp × 10).

**Attribution:** CMEs launch a median of ~7.6h apart, so several CMEs' 4-day windows often contain the *same* storm, and the naive label marks all of them `storm = 1` even though only one physically caused it. To reduce that label noise we also build `storm_attributed`: storm hours are grouped into episodes (gaps > 12h start a new episode), and within each episode only the fastest CME (tie-break: widest half-angle) whose window covers it keeps the positive label; the other candidates are relabeled 0.

In [15]:
import numpy as np

# --- Windowed max Kp per CME (searchsorted: independent overlapping windows) ---
omni_kp = (
    df_omni[['datetime', 'kp_index']]
    .dropna(subset=['kp_index'])
    .sort_values('datetime')
    .reset_index(drop=True)
)
kp_times = omni_kp['datetime'].values
kp_vals  = omni_kp['kp_index'].values  # stored as Kp × 10

t_starts = (flare_matched['cme_time'] + pd.Timedelta('1D')).values
t_ends   = (flare_matched['cme_time'] + pd.Timedelta('4D')).values
idx_starts = np.searchsorted(kp_times, t_starts)
idx_ends   = np.searchsorted(kp_times, t_ends, side='right')

kp_maxes = np.array([
    kp_vals[s:e].max() / 10 if e > s else np.nan
    for s, e in zip(idx_starts, idx_ends)
])

storm_labels = flare_matched[['cme_id']].copy()
storm_labels['kp_max_post'] = kp_maxes
storm_labels['storm'] = (storm_labels['kp_max_post'] >= 5).astype(float)
storm_labels.loc[storm_labels['kp_max_post'].isna(), 'storm'] = np.nan

# --- Storm episodes: consecutive-ish storm hours (Kp >= 5), new episode after a >12h gap ---
storm_hr_mask  = kp_vals >= 50
storm_hr_times = kp_times[storm_hr_mask]
gap_hours   = np.diff(storm_hr_times).astype('timedelta64[h]').astype(int)
episode_ids = np.concatenate([[0], np.cumsum(gap_hours > 12)])
ep_start_time = pd.Series(storm_hr_times).groupby(episode_ids).min()  # episode -> onset
print(f"Storm episodes (Kp>=5, 12h gap rule): {episode_ids.max() + 1}")

# First storm hour inside each CME's window -> the episode this CME's label came from
ep_starts = np.searchsorted(storm_hr_times, t_starts)
ep_ends   = np.searchsorted(storm_hr_times, t_ends, side='right')
storm_labels['episode'] = [
    episode_ids[s] if e > s else -1 for s, e in zip(ep_starts, ep_ends)
]

# --- Attribution: candidates = CMEs whose window covers the episode ---
cand = (
    flare_matched[['cme_id', 'cme_time', 'cme_speed_kms', 'cme_half_angle_deg']]
    .merge(storm_labels.loc[storm_labels['episode'] >= 0, ['cme_id', 'episode']], on='cme_id')
)

# v1: fastest candidate wins (tie-break: widest half-angle)
v1_winner = (
    cand.sort_values(['cme_speed_kms', 'cme_half_angle_deg'], ascending=False, na_position='last')
    .groupby('episode').first()['cme_id']
)

# v2: best arrival-time match. Predicted transit = 1 AU / speed; blame the
# candidate whose predicted arrival is closest to the episode onset. This is
# closer to how DONKI analysts attribute storms than "fastest wins" (no drag
# correction yet — fast CMEs decelerate en route, so this is still approximate).
AU_KM = 1.496e8
cand['pred_arrival'] = cand['cme_time'] + pd.to_timedelta(AU_KM / cand['cme_speed_kms'], unit='s')
cand['timing_err_h'] = (cand['pred_arrival'] - cand['episode'].map(ep_start_time)).abs().dt.total_seconds() / 3600
v2_winner = (
    cand.dropna(subset=['timing_err_h']).sort_values('timing_err_h')
    .groupby('episode').first()['cme_id']
)
v2_winner = v2_winner.reindex(v1_winner.index).fillna(v1_winner)  # no-speed episodes: fall back to v1

for col, winners in [('storm_attributed', v1_winner), ('storm_attributed_v2', v2_winner)]:
    storm_labels[col] = ((storm_labels['episode'] >= 0)
                         & storm_labels['cme_id'].isin(winners.values)).astype(float)
    storm_labels.loc[storm_labels['storm'].isna(), col] = np.nan

labeled = storm_labels['storm'].notna().sum()
print(f"CMEs with post-event Kp data: {labeled} / {len(storm_labels)}")
print(f"Naive label          — storm CMEs: {int((storm_labels['storm'] == 1).sum())} ({(storm_labels['storm'] == 1).mean():.1%})")
print(f"Attributed v1 (speed)— storm CMEs: {int((storm_labels['storm_attributed'] == 1).sum())} ({(storm_labels['storm_attributed'] == 1).mean():.1%})")
print(f"Attributed v2 (time) — storm CMEs: {int((storm_labels['storm_attributed_v2'] == 1).sum())} ({(storm_labels['storm_attributed_v2'] == 1).mean():.1%})")
print(f"v1/v2 pick the same CME for {(v1_winner == v2_winner).mean():.1%} of episodes")

Storm episodes (Kp>=5, 12h gap rule): 158
CMEs with post-event Kp data: 4734 / 4737
Naive label          — storm CMEs: 1079 (22.8%)
Attributed v1 (speed)— storm CMEs: 140 (3.0%)
Attributed v2 (time) — storm CMEs: 140 (3.0%)
v1/v2 pick the same CME for 46.4% of episodes


/Library/Frameworks/Python.framework/Versions/3.14/lib/python3.14/site-packages/pandas/core/arrays/timedeltas.py:1163: RuntimeWarning: invalid value encountered in cast
  int_data = data.astype(np.int64)


### 5. Merge all into final dataset

In [16]:
df_final = (
    flare_matched
    .merge(df_wind_pre, on='cme_id', how='left')
    .merge(storm_labels[['cme_id', 'kp_max_post', 'storm', 'storm_attributed', 'storm_attributed_v2']], on='cme_id', how='left')
)

# CMEs near the end of the dataset won't have a full 4-day post window — drop them
df_final = df_final.dropna(subset=['storm']).copy()
for c in ['storm', 'storm_attributed', 'storm_attributed_v2']:
    df_final[c] = df_final[c].astype(int)

# Flare features stay NaN when no flare matched (has_flare=0): zero-filling would
# invent fake values (e.g. log flux of 0 = 1 W/m² would be an absurdly huge flare).
# XGBoost handles NaN natively; the Ridge pipeline median-imputes with has_flare
# available to disambiguate.

print(df_final.shape)
print(f"\nStorm rate (naive):           {df_final['storm'].mean():.1%}")
print(f"Storm rate (attributed v1):   {df_final['storm_attributed'].mean():.1%}")
print(f"Storm rate (attributed v2):   {df_final['storm_attributed_v2'].mean():.1%}")
print(f"\nMissing value rates:")
print(df_final.isnull().mean()[df_final.isnull().mean() > 0].sort_values(ascending=False))
df_final.head()

(4734, 30)

Storm rate (naive):           22.8%
Storm rate (attributed v1):   3.0%
Storm rate (attributed v2):   3.0%

Missing value rates:
flare_class               0.203633
flare_class_num           0.203633
flare_log_peak_flux       0.203633
flare_duration_min        0.203633
flare_end_time_imputed    0.203633
cme_longitude             0.153570
cme_latitude              0.004225
cme_type_num              0.003380
cme_speed_kms             0.002957
cme_half_angle_deg        0.002957
dtype: float64


,cme_time,cme_speed_kms,cme_half_angle_deg,cme_latitude,cme_longitude,cme_type_num,cme_id,flare_class,flare_class_num,flare_log_peak_flux,...,wind_density_mean,wind_pressure_mean,wind_efield_mean,wind_temp_mean,ctx_f107,ctx_sunspots,kp_max_post,storm,storm_attributed,storm_attributed_v2
0,2017-02-10 13:48:00,594.0,10.0,18.0,148.0,2.0,0,<NA>,NaN,NaN,...,5.500000,2.005278,0.243889,70931.944444,72.027778,22.555556,2.3,0,0,0
1,2017-02-11 06:00:00,456.0,18.0,-1.0,93.0,1.0,1,<NA>,NaN,NaN,...,4.745946,1.616486,0.265946,56423.216216,72.972973,23.486486,2.3,0,0,0
2,2017-02-15 03:09:00,405.0,12.0,16.0,-173.0,1.0,2,<NA>,NaN,NaN,...,12.597222,2.053611,0.010000,18074.472222,72.833333,19.555556,4.3,0,0,0
3,2017-02-18 02:30:00,239.0,17.0,26.0,-110.0,1.0,3,B,1.0,-6.885685,...,6.983333,3.387500,0.509722,149947.583333,73.675000,13.583333,3.7,0,0,0
4,2017-02-18 07:24:00,492.0,27.0,4.0,88.0,1.0,4,<NA>,NaN,NaN,...,6.444444,3.187778,0.266111,137876.500000,73.966667,13.444444,3.7,0,0,0


### 5b. Validation — agreement with NASA analyst attributions

`data/clean-data/donki-cme-features.parquet` carries `leads_to_gst`: CMEs that DONKI analysts explicitly linked to a geomagnetic storm via `linked_events`. Too sparse to train on (silence ≠ innocent), but a high-precision answer key for grading our attribution rules: of the analyst-blamed CMEs, how many does each rule also blame?

In [17]:
donki_cme = pd.read_parquet('data/clean-data/donki-cme-features.parquet')
donki_cme['cme_time'] = pd.to_datetime(donki_cme['start_time'], utc=True).dt.tz_localize(None)

val = df_final.merge(donki_cme[['cme_time', 'leads_to_gst']], on='cme_time', how='left')
val = val.dropna(subset=['leads_to_gst'])
val['leads_to_gst'] = val['leads_to_gst'].astype(int)
linked = val[val['leads_to_gst'] == 1]

print(f"Matched to analyst table: {len(val)} / {len(df_final)}  |  analyst-linked positives: {len(linked)}")
print(f"\nOf {len(linked)} analyst-blamed CMEs:")
print(f"  storm followed at all (naive=1):     {int(linked['storm'].sum())} ({linked['storm'].mean():.1%})  <- window/label sanity check")
print(f"  v1 (fastest wins) picked same CME:   {int(linked['storm_attributed'].sum())} ({linked['storm_attributed'].mean():.1%})")
print(f"  v2 (arrival-time) picked same CME:   {int(linked['storm_attributed_v2'].sum())} ({linked['storm_attributed_v2'].mean():.1%})")

Matched to analyst table: 4726 / 4734  |  analyst-linked positives: 54

Of 54 analyst-blamed CMEs:
  storm followed at all (naive=1):     49 (90.7%)  <- window/label sanity check
  v1 (fastest wins) picked same CME:   13 (24.1%)
  v2 (arrival-time) picked same CME:   13 (24.1%)


In [18]:
df_final.to_csv('data/cme_dataset.csv', index=False)
print("Saved to data/cme_dataset.csv")
print(f"\nFinal columns: {df_final.columns.tolist()}")

Saved to data/cme_dataset.csv

Final columns: ['cme_time', 'cme_speed_kms', 'cme_half_angle_deg', 'cme_latitude', 'cme_longitude', 'cme_type_num', 'cme_id', 'flare_class', 'flare_class_num', 'flare_log_peak_flux', 'flare_duration_min', 'flare_end_time_imputed', 'has_flare', 'wind_bz_mean', 'wind_bz_min', 'wind_bz_south_frac', 'wind_bz_strong_south_frac', 'wind_imf_mean', 'wind_speed_mean', 'wind_speed_max', 'wind_density_mean', 'wind_pressure_mean', 'wind_efield_mean', 'wind_temp_mean', 'ctx_f107', 'ctx_sunspots', 'kp_max_post', 'storm', 'storm_attributed', 'storm_attributed_v2']


---
## Summary of design choices

**What this notebook produces:** `data/cme_dataset.csv` — one row per CME (4,734 rows, 2017-02-09 → 2024-12-31), with features knowable by launch + 12h and three candidate labels. It builds the dataset only; modeling lives elsewhere.

**Choices and why:**

1. **CME-anchored rows** — each row asks "will this CME cause a storm?", preserving the cause→effect time structure that day-binned aggregation destroys.

2. **Time window starts 2017-02-09** (GOES flare catalog start), not 2010. Before 2017, "no flare matched" would mean "no flare *data*", biasing the flare features and entangling them with the solar cycle. Post-2017, NaN flare features genuinely mean "no flare within 6h".

3. **Flares from `data/clean-data/solar-flare-events.parquet`** (team-cleaned: garbage rows dropped, `duration_minutes`/`log_peak_flux`/`end_time_imputed` added), matched by time proximity — nearest flare peaking ≤ 6h before launch. `linked_events` was rejected for matching: only ~12% coverage, and its IDs point to DONKI's own flare entries, not the GOES catalog.

4. **No zero-filling of missing flare features.** They stay NaN with an explicit `has_flare` 0/1 flag. Zero-fill would invent physics (log-flux 0 = 1 W/m², an impossibly large flare). XGBoost handles NaN natively; linear models should impute inside their pipeline (fit on train only).

5. **Solar wind from raw OMNI, not the cleaned parquet** — we need `kp_index` for the label, and NaN-tolerant window aggregation instead of complete-case row dropping (OMNI gaps correlate with extreme conditions). Feature *selection* follows the OMNI EDA: the 5 "strong keeps" + electric field, temperature, and solar-cycle context (F10.7, sunspots).

6. **Wind window = launch − 24h → launch + 12h, via per-CME `searchsorted`.** 83% of CMEs launch within 24h of the previous one, so a `merge_asof` join silently truncates most windows; searchsorted gives every CME its full, independent window. The +12h extension is leakage-free (the CME needs ≥ ~1 day to reach L1; the label window starts at +1 day) and improved attributed-label classification. Cost: decision time is launch + 12h → effective warning 0.5–3.5 days.

7. **Label: max Kp in [launch + 1d, launch + 4d]; storm = Kp ≥ 5** (NOAA G1). Day 0 excluded to avoid flare contamination. Kp ≥ 5 matches the OMNI EDA convention (`kp_index` stored ×10).

8. **Three labels, use deliberately:**
   - `storm` (22.8% positive) — "did a storm *follow*?" Noisy for causal claims: ~7 CMEs share each storm window on average.
   - `storm_attributed` (3.0%) — "did this CME *cause* it?" Storm hours grouped into 158 episodes (>12h gap = new episode); the fastest candidate per episode keeps the label. **Primary attributed label.** Caveat: speed-based winners make speed features partly self-fulfilling.
   - `storm_attributed_v2` (3.0%) — arrival-time variant (1 AU / speed vs episode onset). Kept for comparison; did not improve analyst agreement.

9. **Validation against NASA analysts** (`leads_to_gst`, 54 gold links): 90.7% of analyst-blamed CMEs have `storm = 1` (window machinery is sound), but both attribution rules match the analysts' pick only ~24% of the time. **Known limitation:** attribution from tabular launch data appears capped well below expert level — quote this caveat with any attributed-label results. Analyst links are for validation only; silence ≠ innocent, so never train on `leads_to_gst`.

10. **Modeling guidance:** split train/test chronologically (activity is autocorrelated); report precision/recall/F1/PR-AUC, not accuracy (positives are 3–23%); regression on `kp_max_post` has shown ~no signal — the classification framing is where the signal is.